### PR TITLE
Add DnsResolverConfig::query_retry

### PR DESCRIPTION
**Pull request by jgallagher**
This allows a shorter timeout for re-querying DNS servers when the servers report no records for backends in question.

This came up when running a Nexus test that was sequenced like this:

* Start a DNS server with no records to serve
* Start a qorb pool pointed at that DNS server
* qorb queries the DNS server, sees no records, and goes to sleep for `query_interval`
* A few seconds later, the DNS server is populated with records
* We try to `claim()` a connection from the pool. This times out, because qorb thinks there are no backends (from the initial DNS queries), and we haven't yet hit `query_interval` to check with DNS again

I've confirmed this change fixes that test failure.

### CODE REVIEW

File: src/resolvers/dns.rs [hunk 1]
```diff
@@ -404,11 +434,18 @@ pub struct DnsResolverConfig {
...
+    /// at least one backend record?
     ///
     /// Default: 60 seconds
     pub query_interval: Duration,
 
+    /// How long should we wait before re-querying DNS servers, if we received
+    /// no backend records from a DNS server?
+    ///
+    /// Default: 10 seconds
+    pub query_retry: Duration,
```
Comment (jgallagher) [inline src/resolvers/dns.rs:447]: Bikeshedding welcome on this name.

File: src/resolvers/dns.rs [hunk 2]
```diff
@@ -214,18 +219,35 @@ impl DnsResolverWorker {
...
+        // If we get a `NoRecordsFound` error from a server, we won't consider
+        // the client to have errored, but we also don't want to take that
+        // result as authoritative: if dynamic resolution is in play, we might
+        // have one client that only has records for other DNS servers and other
+        // clients that only have records for backends.
+        //
+        // If at least one client returns `NoRecordsFound` _and_ no clients
+        // return records, we'll return `Some(_)` with an empty list of
+        // backends, to honor what DNS has told us.
+        let saw_no_records_found = Arc::new(AtomicBool::new(false));
```
Comment (jgallagher) [inline src/resolvers/dns.rs:231]: I think this comment is the most important bit - we don't want to treat `NoRecordsFound` as a client _error_, but we also don't want to take it as the first result (particularly due to dynamic resolution).

File: src/resolvers/dns.rs [hunk 3]
```diff
@@ -214,18 +219,35 @@ impl DnsResolverWorker {
         // whether they responded in time or not.
         let first_result = Arc::new(Mutex::new(None));
 
+        // If we get a `NoRecordsFound` error from a server, we won't consider
+        // the client to have errored, but we also don't want to take that
+        // result as authoritative: if dynamic resolution is in play, we might
+        // have one client that only has records for other DNS servers and other
+        // clients that only have records for backends.
```
Comment (smklein) [inline src/resolvers/dns.rs:226]: I'm not sure i follow why this is relevant for dynamic resolution , as opposed to a normal lookup.

It's *always* the case that "some DNS servers may have records that are missing from other DNS servers", independent of whether that backend is for a DNS server or something else, right?

File: src/resolvers/dns.rs (hunk 1 above)
Comment (smklein) [reply src/resolvers/dns.rs:447]: I think this name overlaps too much with `query_interval` - it's more verbose, but I think `query_retry_if_no_records_found` is *really* explicit that this is for an edge-case condition.

File: src/resolvers/dns.rs [hunk 4]
```diff
@@ -587,6 +654,63 @@ mod test {
...
+        catalog: UpdateableCatalog,
+    }
+
+    impl UpdateableServer {
+        async fn remove_backend(&mut self, name: &str) {
+            self.data.remove_backend(name);
+            self.catalog.0.lock().await.upsert(
+                LowerName::from_str(&self.data.domain).unwrap(),
+                self.data.build_authority(),
+            );
```
Comment (smklein) [inline src/resolvers/dns.rs:704]: nit, this could probably be pulled into a little helper and used in both `add_backend` and `remove_backend` since it's identical

File: src/resolvers/dns.rs [hunk 5]
```diff
@@ -878,6 +1002,96 @@ mod test {
...
+            query_interval: Duration::from_secs(60),
+            query_retry: Duration::from_millis(100),
+            ..Default::default()
+        };
+        let mut resolver = DnsResolver::new(service, bootstrap_servers, config);
+        let mut monitor = resolver.monitor();
+
+        // Wait until we've sen the record, then remove it and wait until our
+        // monitor realizes it's gone (via advancing time past
+        // `query_interval`).
```
Comment (smklein) [inline src/resolvers/dns.rs:1077]:
```suggestion
        // Wait until we've seen the record, then remove it and wait until our
        // monitor realizes it's gone (via advancing time past
        // `query_interval`).
```

File: src/resolvers/dns.rs [hunk 6]
```diff
@@ -878,6 +1002,96 @@ mod test {
...
+            .await
+            .expect("successful lookup");
+        assert_eq!(backends.len(), 1);
+    }
+
+    #[tokio::test]
+    async fn test_no_records_found() {
+        setup_tracing_subscriber();
+
+        // Start a DNS server that reports no records for the service.
```
Comment (smklein) [inline src/resolvers/dns.rs:1059]: It's reporting a single record for the service, right? not "no records"

File: src/resolvers/dns.rs [hunk 7]
```diff
@@ -878,6 +1002,96 @@ mod test {
...
+        wait_for_backends(&mut monitor, 0).await;
+
+        // Now put the backend back in place on the DNS server side.
+        dns_server
+            .add_backend(1234, "test001.example.com.", 1000)
+            .await;
+
+        // We should nearly immediately find this new backend, due to the very
+        // low `query_retry` interval.
+        wait_for_backends(&mut monitor, 1).await;
```
Comment (smklein) [inline src/resolvers/dns.rs:1092]: Should we test this with a timeout or something? the comment says "we should nearly immediately" do something, but nothing validates that this happens quickly

**Review by smklein — APPROVED** [review]

File: src/resolvers/dns.rs (hunk 6 above)
Comment (jgallagher) [reply src/resolvers/dns.rs:1059]: Oops, yeah, this is leftover from an earlier version of the test where I did start it with no records. Will fix.

File: src/resolvers/dns.rs (hunk 7 above)
Comment (jgallagher) [reply src/resolvers/dns.rs:1092]: Yep, easy enough.

File: src/resolvers/dns.rs (hunk 3 above)
Comment (jgallagher) [reply src/resolvers/dns.rs:226]: > It's _always_ the case that "some DNS servers may have records that are missing from other DNS servers", independent of whether that backend is for a DNS server or something else, right?

I think this is technically true but not really expected to be true in practice much, right? Or I'll claim something that seems nearly equivalent: it's always the case that any single DNS server's records may change any time, including right after we get query results. But this resolver caches results for _some_ period of time. Prior to this PR, it treats a successful response from any server as "the results to cache", so doesn't attempt to handle different servers having different records. (To be clear, I think this is perfectly reasonable!)

After this PR, if we were to just treat "no records" as authoritative and ignore any other outstanding queries, that would be correct if we expected all the DNS servers to be more or less interchangeable. But for dynamic resolution specifically, this expectation is wrong: _by design_ the servers are not interchangeable, even though we still query them all.

Comment (smklein) [reply src/resolvers/dns.rs:226]: 100% agreed with you, until this specific point

> But for dynamic resolution specifically, this expectation is wrong: _by design_ the servers are not interchangeable, even though we still query them all.

To me, the term "dynamic resolution" means "using DNS to find additional DNS servers".

This is why I'm getting stuck on the use of "dynamic resolution" being a motivator. If we spun up a qorb pool with three hard-coded DNS server addresses, we would have **no dynamic resolution**. However, we could still see different results from different DNS servers (maybe they're getting concurrently populated while we're querying them?). Thus, as you say, they still aren't completely interchangeable, even thought we're statically accessing DNS.

Comment (jgallagher) [reply src/resolvers/dns.rs:226]: I think I'll claim that we're already broken in that case, then? If those three servers return different records for the backend we're querying, we're only taking the first to respond and ignoring the others.

Peeling back the curtain a bit: I made the changes to this function last; without them, the new stuff works fine for all the existing qorb tests, _except_ the dynamic resolution test, which needed these changes. That definitely motivated the comment!

Comment (smklein) [reply src/resolvers/dns.rs:226]: This may be a sign of lack of sufficient coverage rather than an argument for it being relevant only to dynamic resolution.

Both "looking up DNS servers for dynamic resolution" and "looking up backends, with or without dynamic resolution" use this `dns_lookup` function.

I'm happy with this behavior you're adding, I just think it applies to **all** lookups, and whether or not it's being used to find "DNS servers" or "any other server" is orthogonal.

Comment (jgallagher) [reply src/resolvers/dns.rs:226]: I think we basically agree on all of this, haha. I'll take a shot at rewording the comment.

**Review by smklein — APPROVED** [review]
Looks great, thanks for the updates!
